---
title: "SystemPipeR Exploration"
author: "Angelo Pelonero"
date: "9/6/2019"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Intro:

This is an appropriation of a [UCR RNA-seq workflow](http://girke.bioinformatics.ucr.edu/GEN242/pages/mydoc/systemPipeRNAseq.html) using systemPipeR. To quote the introduction to the linked vignette:

> To run this sample report, mini sample FASTQ and reference genome files can be downloaded from here. The chosen data set SRP010938 contains 18 paired-end (PE) read sets from Arabidposis thaliana (Howard et al. 2013). To minimize processing time during testing, each FASTQ file has been subsetted to 90,000-100,000 randomly sampled PE reads that map to the first 100,000 nucleotides of each chromosome of the A. thalina genome. The corresponding reference genome sequence (FASTA) and its GFF annotion files (provided in the same download) have been truncated accordingly. This way the entire test sample data set is less than 200MB in storage space. A PE read set has been chosen for this test data set for flexibility, because it can be used for testing both types of analysis routines requiring either SE (single end) reads or PE reads.

### load packages
```{r}
library(systemPipeR)
library(systemPipeRdata)
genWorkenvir(workflow="rnaseq")
```
DEV NOTE: "genWorkenvir(workflow="rnaseq")" generated an rnaseq directory. We need to look into exactly what this does and why before recommending systemPipeR for analysis (versus running a standard pipeline more simiar to Rebecca's preprocessing workflow)

>hopefully won't need this, but will leave for now: "If applicable load custom functions not provided by"
```{r source_helper_fcts, eval=FALSE}
source("systemPipeRNAseq_Fct.R")
```

### Define the experiments for systemPipeR
DEV NOTE: Experiment definition provided by `targets` file

The `targets` file defines all FASTQ files and sample comparisons of the analysis workflow.
```{r load_targets, eval=TRUE}
targetspath <- system.file("extdata", "targets.txt", package="systemPipeR")
targets <- read.delim(targetspath, comment.char = "#")[,1:4]
targets
```

# Read preprocessing

##$ Read quality filtering and trimming

The function `preprocessReads` allows to apply predefined or custom read preprocessing functions to all FASTQ files referenced in a `SYSargs` container, such as quality filtering or adaptor trimming routines. The following example performs adaptor trimming with the `trimLRPatterns` function from the `Biostrings` package. After the trimming step a new targets file is generated (here `targets_trim.txt`) containing the paths to the trimmed FASTQ files. The new targets file can be used for the next workflow step with an updated `SYSargs` instance, _e.g._ running the NGS alignments using the trimmed FASTQ files.
```{r fastq_filter, eval=FALSE}
args <- systemArgs(sysma="param/trim.param", mytargets="targets.txt")
preprocessReads(args=args, Fct="trimLRPatterns(Rpattern='GCCCGGGTAA', subject=fq)",
                batchsize=100000, overwrite=TRUE, compress=TRUE)
writeTargetsout(x=args, file="targets_trim.txt", overwrite=TRUE)
```

### FASTQ quality report

The following `seeFastq` and `seeFastqPlot` functions generate and plot a series of useful 
quality statistics for a set of FASTQ files including per cycle quality box
plots, base proportions, base-level quality trends, relative k-mer
diversity, length and occurrence distribution of reads, number of reads
above quality cutoffs and mean quality distribution. The results are
written to a PDF file named `fastqReport.pdf`.

```{r fastq_report, eval=FALSE}
args <- systemArgs(sysma="param/tophat.param", mytargets="targets.txt")
fqlist <- seeFastq(fastq=infile1(args), batchsize=100000, klength=8)
pdf("./results/fastqReport.pdf", height=18, width=4*length(fqlist))
seeFastqPlot(fqlist)
dev.off()
``` 

DEV NOTE: may be best to view plot in preview, but play around with displaying plot in RStudio.

### Alignments

DEV NOTE: This vignette uses a compute cluster to run bowtie2/tophat2. We may be better off showing how to run locally - pause here for now.

